mergebot_poller.py

```diff
@@ -1,47 +1,179 @@
-"""TODO(jasonkuster): DO NOT SUBMIT without one-line documentation for mergebot_poller.
+"""mergebot_poller is a utility which polls an SCM and follows commands.
 
-TODO(jasonkuster): DO NOT SUBMIT without a detailed description of mergebot_poller.
+mergebot_poller defines a MergebotPoller class from which other classes can
+inherit and a create_poller method which allows creation of SCM pollers.
 """
 
-import mergebot
+
 import Queue
+import sys
+from threading import Thread
+import time
+import github_helper
+import merge
+
+# TODO(jasonkuster): Fetch authorized users from somewhere official.
+AUTHORIZED_USERS = ['davorbonaci']
+BOT_NAME = 'merge-bot'
+ORG = 'apache'
+
+
+def create_poller(config):
+    """create_poller creates a poller of the type specified in the config.
+
+    Args:
+        config: A dictionary containing repository configuration.
+    Returns:
+        MergebotPoller of type specified in configuration
+    Raises:
+        AttributeError: if passed an unsupported SCM type.
+    """
+    if config['scm_type'] == 'github':
+        return GithubPoller(config)
+    raise AttributeError('Unsupported SCM Type: {}.'.format(config['scm_type']))
+
+
+class MergebotPoller(object):
+    """MergebotPoller is a base class for polling SCMs.
+    """
+
+    def __init__(self, config):
+        self.config = config
+        # Instantiate the two variables used for tracking work.
+        # Thread-safe because of the GIL.
+        self.work_queue = Queue.Queue()
+        self.known_work = {}
+
+    def poll(self):
+        """Poll should be implemented by subclasses as the main entry point.
+
+        Raises:
+            NotImplementedError: Must be implemented by subclasses.
+        """
+        raise NotImplementedError()
+
+
+class GithubPoller(MergebotPoller):
+    """GithubPoller polls Github repositories and merges them.
+    """
+
+    def __init__(self, config):
+        super(GithubPoller, self).__init__(config)
+        self.COMMANDS = {'merge': self.merge_git}
+        # Set up a github helper for handling network requests, etc.
+        self.github_helper = github_helper.GithubHelper(ORG, self.config[
+            'repository'])
+
+    def poll(self):
+        """Kicks off polling of Github.
+        """
+        _print_flush('Starting merge poller for {}.'.format(self.config[
+                                                                'name']))
+        self.poll_github()
+
+    def poll_github(self):
+        """Polls Github for PRs, verifies, then searches them for commands.
+        """
+        # Kick off an SCM merger
+        merger_args = (self.work_queue, self.known_work, self.github_helper,)
+        merger = Thread(target=merge.merge_git, args=merger_args)
+        merger.start()
+        # Loop: Forever, every fifteen seconds.
+        while True:
+            # Load list of pull requests from Github
+            prs = self.github_helper.fetch_prs()
+            if prs:
```

Comment:
Since this always returns a list, I think we can just do:

```
# Loop: Forever, with 15 seconds in between iterations
while True:
  for pull in self.github_helper.fetch_prs():
    self.check_pr(pull)
  time.sleep(15)
```